price-feeder/config/currency_provider_tracker.go

```diff
@@ -111,51 +117,69 @@ func (t *CurrencyProviderTracker) setCoinIDSymbolMap() error {
 	}
 
 	for _, coin := range listResponse {
-		t.coinIDSymbolMap[coin.Symbol] = coin.ID
+		if _, ok := t.coinIDSymbolMap[coin.Symbol]; !ok {
+			t.coinIDSymbolMap[coin.Symbol] = coin.ID
+		}
 	}
 
 	return nil
 }
 
-// setCurrencyProviders queries CoinGecko's tickers endpoint and the osmosis-api assetpairs
-// endpoint to get all the exchanges that support each price feeder currency pair and store
-// it in the CurrencyProviders map.
-func (t *CurrencyProviderTracker) setCurrencyProviders() error {
-	for _, pair := range t.pairs {
-		// check if CoinGecko API supports pair
-		pairBaseID := t.coinIDSymbolMap[strings.ToLower(pair.Base)]
-		coinGeckoResp, err := http.Get(fmt.Sprintf("%s/%s/%s", coinGeckoRestURL, pairBaseID, coinGeckoTickersEndpoint))
-		if err != nil {
-			return err
-		}
-		defer coinGeckoResp.Body.Close()
+// setOsmosisV2CurrencyProviders queries the osmosis-api assetpairs endpoint to get
+func (t *CurrencyProviderTracker) setOsmosisAPIPairs() error {
+	client := &http.Client{
+		Timeout: requestTimeout,
+	}
+	t.osmosisAPIPairs = make(map[string]string)
 
-		var tickerResponse coinTickerResponse
-		if err = json.NewDecoder(coinGeckoResp.Body).Decode(&tickerResponse); err != nil {
+	osmosisResp, err := client.Get(fmt.Sprintf("%s/%s", osmosisV2RestURL, osmosisV2AssetPairsEndpoint))
+	if err != nil {
+		t.logger.Error().Err(err).Msg("Failed to query osmosis api assetpairs endpoint")
+	} else {
+		defer osmosisResp.Body.Close()
+		var assetPairsResponse []assetPair
+		if err = json.NewDecoder(osmosisResp.Body).Decode(&assetPairsResponse); err != nil {
 			return err
 		}
 
-		for _, ticker := range tickerResponse.Tickers {
-			if ticker.Target == pair.Quote {
-				t.CurrencyProviders[pair.Base] = append(t.CurrencyProviders[pair.Base], ticker.Market.Name)
-			}
+		for _, assetPair := range assetPairsResponse {
+			t.osmosisAPIPairs[assetPair.Base] = assetPair.Quote
 		}
+	}
 
-		// check if osmosis-api supports pair
-		osmosisResp, err := http.Get(fmt.Sprintf("%s/%s", osmosisV2RestURL, osmosisV2AssetPairsEndpoint))
-		if err != nil {
-			return err
-		}
-		defer osmosisResp.Body.Close()
+	return nil
+}
 
-		var assetPairsResponse []assetPair
-		if err = json.NewDecoder(osmosisResp.Body).Decode(&assetPairsResponse); err != nil {
-			return err
+// setCurrencyProviders queries CoinGecko's tickers endpoint to get all the exchanges
+// that support each price feeder currency pair and store it in the CurrencyProviders map.
+func (t *CurrencyProviderTracker) setCurrencyProviders() error {
+	client := &http.Client{
+		Timeout: requestTimeout,
+	}
+	for _, pair := range t.pairs {
+		t.CurrencyProviders[pair.Base] = []string{}
+
+		// check if its a pair supported by the osmosis api
+		if t.osmosisAPIPairs[strings.ToUpper(pair.Base)] == strings.ToUpper(pair.Quote) {
+			t.CurrencyProviders[pair.Base] = append(t.CurrencyProviders[pair.Base], "osmosisv2")
```

Comment:
I'm not sure about the flow you want to get here.
1. firstly we set currency providers here based on coinGeck data
2. but we also have osmosis provider, which we set here and in the function above. Why in two places? I would update `CurrencyProviders` when calling `setOsmosisAPIPairs` rather than making this weird relationship. Other idea would be to provide the osmosis pairs as the function argument, rather than keeping them in the struct.

Reply:
Sending in the osmosis pairs as a function arg now👍